### PR TITLE
rocksdb: Update to 5.0 and use DeleteRange for expirations

### PR DESCRIPTION
**Pull request by theopolis**
This updates RocksDB to v5.1.4 on Linux/macOS and introduces `DeleteRange` for expirations in events. This allows us to bulk-remove event data when events are buffered beyond the configured max.

This also disables the WAL for writes to events and forces a flush when the database is closed.

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit fe0d72460188d347422464dc785a29d37ae00fa5 (Job results: [1188](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1188/)) failed one or more tests (**Windows**).

File: tools/provision/formula/rocksdb.rb [hunk 1]
```diff
@@ -3,8 +3,8 @@
 class Rocksdb < AbstractOsqueryFormula
   desc "Persistent key-value store for fast storage environments"
   homepage "http://rocksdb.org"
-  url "https://github.com/facebook/rocksdb/archive/v4.11.2.tar.gz"
-  sha256 "9374be06fdfccbbdbc60de90b72b5db7040e1bc4e12532e4c67aaec8181b45be"
+  url "https://github.com/facebook/rocksdb/archive/v5.1.4.tar.gz"
```
Comment (theopolis) [inline tools/provision/formula/rocksdb.rb]: @poppyseedplehzr when you have a second later this week, can you bump Window's RocksDB to 5.1.4?

**Comment by osqueryer** [comment]
:-1: The commit 7058a676804ec3cc69783c4fb04575020f08a8ac (Job results: [1189](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1189/)) failed one or more tests (**Windows**).

**Comment by poppyseedplehzr** [comment]
If you don't mind holding off on bringing this in, I could probably get the Windows rocksdb package updated rq.

File: tools/provision/formula/rocksdb.rb (hunk 1 above)
Comment (poppyseedplehzr) [inline tools/provision/formula/rocksdb.rb]: Hah, saw this after my comment. Will do!

**Comment by poppyseedplehzr** [comment]
If you rebase this, we should be good to test again.

File: osquery/database/plugins/ephemeral.cpp [hunk 2]
```diff
@@ -76,6 +80,22 @@ Status EphemeralDatabasePlugin::remove(const std::string& domain,
   return Status(0);
 }
 
+Status EphemeralDatabasePlugin::removeRange(const std::string& domain,
+                                            const std::string& low,
+                                            const std::string& high) {
+  std::vector<std::string> keys;
+  for (const auto& it : db_[domain]) {
```
Comment (poppyseedplehzr) [inline osquery/database/plugins/ephemeral.cpp]: Does `DBType db_` here have any issue if, in the off chance, `domain` gets set to `""` above?

File: osquery/database/plugins/sqlite.cpp [hunk 3]
```diff
@@ -256,6 +261,31 @@ Status SQLiteDatabasePlugin::remove(const std::string& domain,
...
 
+Status SQLiteDatabasePlugin::removeRange(const std::string& domain,
+                                         const std::string& low,
+                                         const std::string& high) {
+  if (read_only_) {
+    return Status(0, "Database in readonly mode");
+  }
+
+  sqlite3_stmt* stmt = nullptr;
+  std::string q = "delete from " + domain + " where key >= ?1 and key <= ?2;";
```
Comment (poppyseedplehzr) [inline osquery/database/plugins/sqlite.cpp]: Same q here, looks like things will be fine with `domain = ""`, but is there any concern about logging or notifying the user?

File: osquery/events/events.cpp [hunk 4]
```diff
@@ -289,24 +288,46 @@ void EventSubscriberPlugin::expireCheck(bool cleanup) {
...
+    threshold_key = boost::lexical_cast<size_t>(last_key) - getEventsMax();
+
+    // Scan each of the keys in keys, if their ID portion is < threshold.
+    // Nix them, this requires lots of conversions, use with care.
+    std::string max_key;
+    std::string min_key;
+    size_t min_key_value = 0;
+    size_t max_key_value = 0;
+    for (const auto& key : keys) {
+      size_t key_value = std::stoul(key.substr(key.rfind('.') + 1));
```
Comment (poppyseedplehzr) [inline osquery/events/events.cpp]: Would this be better done with `osquery::safeStrtol`?

**Comment by osqueryer** [comment]
:-1: The commit be10d71fbea1d67de80d2ac23de123f6d74f28ad (Job results: [1205](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1205/)) failed one or more tests (**Windows**).

**Comment by poppyseedplehzr** [comment]
Ok to test

**Comment by osqueryer** [comment]
:-1: The commit be10d71fbea1d67de80d2ac23de123f6d74f28ad (Job results: [1206](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1206/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:no_good: The commit 7e69bc2a596b44ec398d0399bbc93a9c2429f065 (Job results: [1194](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1194/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 7e69bc2a596b44ec398d0399bbc93a9c2429f065 (Job results: [1208](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1208/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 7e69bc2a596b44ec398d0399bbc93a9c2429f065 (Job results: [1209](https://jenkins.osquery.io/job/osqueryPullRequestWindows/1209/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 7e69bc2a596b44ec398d0399bbc93a9c2429f065 (Job results: [4275](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4275/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 7e69bc2a596b44ec398d0399bbc93a9c2429f065 (Job results: [4276](https://jenkins.osquery.io/job/osqueryPullRequestLinux/4276/)) failed one or more tests (**Linux**).

**Pull request merged**